### PR TITLE
[Feat][GLM5.2] Shared KV Cache for Prefill CP

### PR DESCRIPTION
**Pull request by taoyuanyuan**
## Motivation

DSA Prefill context parallelism requires every CP rank to consume the complete Main KV and Indexer history. Upstream main supports a layer-sharded path through [#29421](https://github.com/sgl-project/sglang/pull/29421), using NCCL and remaining applicable across NVLink domains. This PR adds another option for same-host deployments within one peer-accessible NVLink domain: CP ranks share one CUDA VMM view and directly read peer-owned HBM pages. The two paths cover different topology and communication trade-offs.

Like layer sharding, the shared layout stores one aggregate Main-K/Indexer cache instead of eight replicated copies, providing up to 8x logical cache capacity over Base CP8 for the same KV memory budget. The difference is the access model: instead of materializing each owner layer on every rank through a collective, attention kernels directly read peer-owned pages through the shared VMM view. On full GLM-5.2-FP8 / H20-3e, Shared KV stays within 1.86% of Base CP8 warmed mean Prefill latency through 256K. Base CP8 fits about 380K tokens in the tested configuration; Shared KV is intentionally capped at 1.2M tokens for the benchmark and successfully serves a 1M input. The tested 100K L2 host-cache restore reduces latency by 92.67%.

## Modifications

- Add `--enable-dsa-shared-kv-cache` for same-host DSA L1/L2 sharing.
- Store Main KV and Indexer once per CP group with owner-sharded CUDA VMM pages; kernels read the group-wide cache directly without per-layer KV broadcast.
- Preserve the same ownership in page-first HiCache and Mooncake Prefill transfer, with mandatory CP synchronization before peer data is consumed.

## Design

Logical page `p` is owned by CP rank `p % cp_size`, and each rank allocates only its owner pages. CUDA VMM exposes the physical allocations as one group-wide tensor view. Main KV uses the canonical view; Indexer adds a rank-local alias over the same physical allocation for the paged DeepGEMM path, without duplicating HBM. A mandatory CP write fence runs before any rank reads peer-owned pages.

Page-first L2 keeps the same logical IDs and ownership. Each rank allocates pinned DRAM only for its owner pages and transfers only between local DRAM and local HBM. CPU failure agreement and a device fence complete each group-wide backup or restore. Mooncake Prefill transfer uses bounded registered staging buffers rather than a second HBM cache.

The current scope is one host, one peer-accessible NVLink domain, GLM-5.2 DSA, FP8 KV, `tp=attn_cp`, interleaved Prefill CP, `pp=1`, page-first L1/L2, and Mooncake Prefill transfer. HiSparse, speculative decoding, and L3 storage are not enabled with this path.

## Trade-offs

Both shared KV and layer sharding remove CP cache replication. They differ in how each rank accesses the complete cache:

**Layer sharding**

```mermaid
flowchart LR
    L1["Layers are sharded across CP ranks"]
    L2["Owner rank provides the current layer KV"]
    L3["Per-layer NCCL collective"]
    L4["Materialize full-layer receive scratch<br/>on every CP rank"]
    L5["Attention reads local HBM"]
    LC["General NCCL-connected topology<br/>Cost: transfer, collective sync, full-layer scratch"]
    L1 --> L2 --> L3 --> L4 --> L5 --> LC
```

**Shared KV (this PR)**

```mermaid
flowchart LR
    S1["Logical page p → owner rank p mod CP size"]
    S2["Owner-sharded HBM mapped into<br/>one group-wide CUDA VMM view"]
    S3["Owner-only page write"]
    S4["Mandatory CP publication fence"]
    S5["Attention directly reads<br/>local or peer-owned pages"]
    SC["One peer-accessible NVLink domain<br/>No per-layer KV collective or full-layer scratch"]
    S1 --> S2 --> S3 --> S4 --> S5 --> SC
```

| | Layer sharding | Shared KV |
| --- | --- | --- |
| Data access | Materialize the owner layer, then read local HBM | Read peer-owned pages through one VMM address space |
| Communication | Per-layer collective transfer | No per-layer KV collective; mandatory write fence |
| Temporary HBM | Full-layer receive scratch | No full-layer scratch |
| Topology | Works across general NCCL-connected topologies | Requires one peer-accessible NVLink domain |
| Main cost | Data movement and collective synchronization | Peer-HBM latency and write synchronization |

In the current upstream implementation, the layer-sharded path has not yet optimized its communication to fetch only active/sparse KV. Under the matched setup used here, Shared KV reduces warmed mean Prefill latency by 49.01% at 4K (798.09 vs. 406.91 ms), 16.26% at 128K (14,877.29 vs. 12,457.69 ms), and 7.87% at 1M (179,701.77 vs. 165,565.18 ms). These results describe the current implementations rather than an inherent limit of layer sharding; the comparison should be repeated after its sparse-fetch communication path is optimized.

### L1 Prefill

Hardware/model: one H20-3e node, full 78-layer GLM-5.2-FP8, TP8/CP8, FP8 KV, `page_size=64`, `mem_fraction_static=0.88`, FlashMLA Sparse Prefill, FlashMLA KV Decode, and 8192-token chunks. Base CP8 disables cache sharding and supports about 380K tokens in this configuration. Shared KV uses `--max-total-tokens 1200000` as a test limit; this configured limit is not the architectural capacity ratio. Base CP8 uses `--cuda-graph-max-bs-decode 24` to fit its maximum capacity. All other workload and backend arguments match.

**Metric:** mean end-to-end latency over three measured requests after one warmup request and a cache flush. This is a mean, not P90. With output length 1, E2E latency is used as Prefill latency. Actual concurrency is 0.995-1.000 and cache hit rate is 0% for every row.

| Input | Base CP8 | Shared KV | Shared vs Base CP8 |
| ---: | ---: | ---: | ---: |
| 4K | 399.87 ms | 406.91 ms | +1.76% |
| 32K | 2,891.98 ms | 2,913.81 ms | +0.75% |
| 64K | 5,869.68 ms | 5,925.54 ms | +0.95% |
| 128K | 12,331.04 ms | 12,457.69 ms | +1.03% |
| 256K | 27,087.90 ms | 27,591.12 ms | +1.86% |
| 512K | N/A (capacity) | 66,010.15 ms | N/A |
| 1M | N/A (capacity) | 165,565.18 ms | N/A |

The 4K-to-256K rows are matched performance comparisons. The 512K and 1M rows demonstrate the additional capacity and are not Base CP8 latency comparisons.

<details>
<summary>Server and benchmark commands</summary>

```bash
SGLANG_JIT_DEEPGEMM_PRECOMPILE=0 \
PYTHONPATH=$PWD/python:$PWD/sgl-kernel/python \
python3 -m sglang.launch_server \
  --model-path /home/models/GLM-5.2-FP8 \
  --tokenizer-path /home/models/GLM-5.2-FP8 \
  --trust-remote-code \
  --host 0.0.0.0 --port 30080 \
  --tp-size 8 --attn-cp-size 8 \
  --attention-backend dsa \
  --dsa-prefill-backend flashmla_sparse \
  --dsa-decode-backend flashmla_kv \
  --enable-prefill-cp --cp-strategy interleave \
  --kv-cache-dtype fp8_e4m3 \
  --page-size 64 --mem-fraction-static 0.88 \
  --max-total-tokens 1200000 \
  --chunked-prefill-size 8192 \
  --skip-server-warmup \
  --enable-dsa-shared-kv-cache

python3 -m sglang.benchmark.serving \
  --backend sglang --host 127.0.0.1 --port 30080 \
  --dataset-name random-ids \
  --model /home/models/GLM-5.2-FP8 \
  --tokenizer /home/models/GLM-5.2-FP8 \
  --num-prompts 3 --max-concurrency 1 \
  --random-input-len ${INPUT_LEN} --random-output-len 1 \
  --random-range-ratio 1.0 --request-rate inf \
  --tokenize-prompt --warmup-requests 1 --flush-cache \
  --cache-report --output-details --disable-tqdm \
  --seed ${SEED} --output-file ${OUTPUT_FILE}
```

</details>

### L2 HiCache

With a 150K-token L1, `hicache_ratio=10`, page-first layout, `write_through`, kernel IO, and no L3, each rank allocates 11.57 GB for its owner pages. CP8 provides one aggregate 92.56 GB / 23,528-page L2 cache instead of eight replicated logical copies.

An A/B1/B2/A test uses three distinct 100K prompts so B1 and B2 evict A from L1. Restoring A reports `device=0, host=99968`, reproduces `Paris` / token `[12089]`, and reduces E2E latency from 10,509.29 ms to 770.31 ms (-92.67%). No L3 backend is configured or observed.

## Correctness

- Fixed short, 100K, and 128K output-1 prompts match Base CP8; a short semantic decode also matches.
- GSM8K500 24-shot completion scores `0.956` against the registered `0.94` threshold. MMLU64 sanity scores `0.796875`, equal to ordinary Base.
- Targeted CPU tests pass `90` cases, CP8 multi-GPU tests pass `3` cases, and the L2 host-only restore reproduces the cold answer.

All changed-file pre-commit hooks and `git diff --check` pass.

Co-authored with [@foraxe](https://github.com/foraxe)

---
### CI States

Latest PR Test (Base): :no_entry_sign: [Run #30150164674](https://github.com/sgl-project/sglang/actions/runs/30150164674)
Latest PR Test (Extra): :x: [Run #30150164598](https://github.com/sgl-project/sglang/actions/runs/30150164598)

### CODE REVIEW

**Comment by yiminghub2024** [comment]
@taoyuanyuan  good feature , i need it , but i have two questions . 
1. which l2 kvcache product you use ? hicache or mooncake or just hbm ?
2. if your feature can use together with --enable-dsa-cache-layer-split and --enable-dsa-shared-kv-cache?
3. both two pr purpers is for more kvcache ? or kvcache not changed ,only reduce ttft ?
4. why --enable-dsa-cache-layer-split merged many days ,but docker images for 0.5.15.post1 not include ,i run it says:"no command --enable-dsa-cache-layer-split" 
5. if this pr support   speculative decoding include despark ?

**Comment by taoyuanyuan** [comment]
> @taoyuanyuan good feature , i need it , but i have two questions .
> 
> 1. which l2 kvcache product you use ? hicache or mooncake or just hbm ?
> 2. if your feature can use together with --enable-dsa-cache-layer-split and --enable-dsa-shared-kv-cache?
> 3. both two pr purpers is for more kvcache ? or kvcache not changed ,only reduce ttft ?
> 4. why --enable-dsa-cache-layer-split merged many days ,but docker images for 0.5.15.post1 not include ,i run it says:"no command --enable-dsa-cache-layer-split"
> 5. if this pr support   speculative decoding include despark ?

@yiminghub2024  Thanks for your interest:
1. Cache hierarchy: L1 is the shared HBM KV/indexer cache implemented with CUDA VMM. L2 uses SGLang HiCache in pinned host DRAM. Mooncake is used only as the PD transfer backend, not as the L2 cache. L3 storage is not supported yet, but the owner-sharded page design can be extended to L3 in future work.
2. Feature flags: --enable-dsa-cache-layer-split and --enable-dsa-shared-kv-cache are alternative cache layouts and cannot be enabled together. Please select one of them.
3. Purpose: Both approaches reduce KV/indexer replication across Prefill CP ranks, allowing the same HBM capacity to hold more cache tokens, longer contexts, or more concurrent requests. They do not change the logical KV content. The lower TTFT is an additional performance benefit observed with the shared-VMM path in our workloads.
5. Speculative decoding: This PR does not currently support speculative decoding, including EAGLE and MTP. These paths require separate correctness and performance work.

**Comment by yiminghub2024** [comment]
so  --enable-dsa-cache-layer-split only need l1  and --enable-dsa-shared-kv-cache need l2 hicache?  which is better of them two ? @taoyuanyuan

**Comment by taoyuanyuan** [comment]
@yiminghub2024 Both are L1 KV-cache sharding methods. Shared KV does not require L2 HiCache; L2 is optional.
On a same-host NVLink system, Shared KV is faster in our tests. LayerSplit is better for more general topologies. Both increase KV-cache capacity.

File: python/sglang/srt/mem_cache/dsa_shared_host.py [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+        PoolEntry(
+            name=PoolName.INDEXER,
+            host_pool=index_host,
+            device_pool=kv_pool,
+            layer_mapper=layer_mapper,
+        ),
+    ]
+    device_sync_group = dist.new_group(
+        ranks=dist.get_process_group_ranks(attn_cp_group), backend="nccl"
+    )
```
[the remark this answers is not in the document]
Comment (taoyuanyuan) [reply python/sglang/srt/mem_cache/dsa_shared_host.py:375]: false positive

File: python/sglang/srt/mem_cache/dsa_shared_host.py [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+        self.logical_allocator = _SharedDSAHostPageAllocator(
+            local_page_num=self.anchor_entry.host_pool.page_num,
+            page_size=self.page_size,
+            cp_size=shared_size,
+        )
+        self.size = self.logical_allocator.global_page_num * self.page_size
+        self._sync_flag = torch.ones(1, dtype=torch.int32, device=sync_device)
+
+    def _synchronize(self, error):
+        success = all_ranks_ok(self.cpu_sync_group, error is None)
```
[the remark this answers is not in the document]
Comment (taoyuanyuan) [reply python/sglang/srt/mem_cache/dsa_shared_host.py:146]: false positive

File: python/sglang/srt/mem_cache/dsa_shared_host.py [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+    storage_backend_extra_config=None,
+    enable_storage_metrics=False,
+):
+    if attn_cp_group is None:
+        raise ValueError("DSA shared HiCache requires an attention CP group.")
+    if storage_backend is not None:
+        raise ValueError("DSA shared HiCache does not support L3 storage yet.")
+
+    shared_rank = dist.get_rank(attn_cp_group)
+    shared_size = dist.get_world_size(attn_cp_group)
```
[the remark this answers is not in the document]
Comment (taoyuanyuan) [reply python/sglang/srt/mem_cache/dsa_shared_host.py:331]: For the three comments in dsa_shared_host.py, attn_cp_group is already a raw
Gloo ProcessGroup here. CacheInitParams.attn_cp_cache_group is populated from
attn_cp_cpu_group in kv_cache_builder.py before reaching this builder.
Therefore these torch.distributed calls are valid. The CP8 L2 E2E restore test
also exercises this path successfully.

File: python/sglang/srt/mem_cache/dsa_cache_shared.py [hunk 4]
```diff
@@ -0,0 +1,943 @@
...
+                            aligned_bytes,
+                            [make_rw_access_desc(device_id)],
+                            1,
+                        ),
+                        f"cuMemSetAccess(rank_local={peer_rank})",
+                    )
+        except BaseException as e:
+            mapping_error = e
+
+        _synchronize_vmm_stage(cpu_group, rank, "mapping", mapping_error)
```
[the remark this answers is not in the document]
Comment (taoyuanyuan) [reply python/sglang/srt/mem_cache/dsa_cache_shared.py]: Fixed in 2695507ade. POSIX FD socket setup now completes a
rank-wide error agreement before any rank enters the path exchange.
Failures from directory creation, socket creation, bind, or listen are
propagated consistently to every rank. I also added fault-injection tests
for both local and remote setup failures and verified the real CP8 POSIX
fallback path.

**Comment by taoyuanyuan** [comment]
Replaced by https://github.com/sgl-project/sglang/pull/35309
